### ISSUE TITLE
[Proposal] Add device-awareness to cache and hook layers for efficient tensor handling

### ISSUE THREAD

**Issue by jbloomAus** [neelnanda-io/TransformerLens#246]
### Proposal 

Implement a feature that allows cache to keep tensors on device and enables hook layers to be aware of their devices in the run_with_cache function.

### Motivation

The current implementation does not provide a way for cache to keep tensors on device or for hook layers to be aware of their devices. This can lead to performance issues, as communicating large tensors between devices can significantly impact performance. Addressing this concern will improve the efficiency of the run_with_cache function.

### Pitch

Introduce a self.device parameter for each HookPoint, which will be used by the run_with_cache function to determine the device to store tensors on, unless overridden by the user. This will allow tensors to be kept on their appropriate devices and reduce communication overhead, leading to improved performance.

### Alternatives

Implement an automatic device-aware caching mechanism that detects the device being used and keeps tensors on the appropriate device without any user intervention. This solution might be more complex but would provide a seamless experience for users.

Instead of modifying the run_with_cache function, create a new function specifically designed to handle device-aware caching. This would leave the original function untouched and provide users with the option to use either the original or the new device-aware caching function.

### Additional context

This is a follow-up feature to: https://github.com/neelnanda-io/TransformerLens/pull/125, which added support for having a model with layers split across several GPUs. 

When this feature is completed, we will review https://github.com/neelnanda-io/TransformerLens/issues/105 and possibly complete it. 

### Checklist

- [x] I have checked that there is no similar [issue](https://github.com/Farama-Foundation/MiniGrid/issues) in the repo (**required**)

**Comment by jbloomAus** [comment]
Fixing this card will likely sort out https://github.com/neelnanda-io/TransformerLens/issues/217, so we should check before this card is closed that  this code block no longer shows the cache sending everything to the wrong device. 
```
model = HookedTransformer.from_pretrained(model_name, device="cuda:2")
token = model.to_tokens("testing 1 2 3")
logits, cache = model.run_with_cache(token)
print(logits.get_device())
print(model.cfg.device)
print(cache["blocks.0.mlp.hook_post"].get_device())
```

**Comment by jbloomAus** [comment]
We should also ensure that this issue (https://github.com/neelnanda-io/TransformerLens/issues/202) is solved when we solve this card just because it's fairly relevant.

**Comment by jbloomAus** [comment]
Thanks @slavachalnev  for solving this!! (solved by #250 )
